### PR TITLE
Implement Fuse with Option

### PR DESCRIPTION
**Pull request by cuviper**
The former `done` flag was roughly similar to an `Option` tag, but left
the possibity of misuse. By using a real `Option`, we can set `None`
when the iterator is exhausted, removing any way to call it again. We
also allow niche layout this way, so the `Fuse` may be smaller.

The `FusedIterator` specialization does want to ignore the possibility
of exhaustion though, so it uses `unsafe { intrinsics::unreachable() }`
to optimize that branch away. The entire `Fuse` implementation is now
isolated in its own module to contain that unsafety.

r? @scottmcm

### CODE REVIEW

**Comment by cuviper** [comment]
This was extracted from #70332 -- let's go ahead and get a perf run:

@bors try @rust-timer queue

**Comment by cuviper** [comment]
If this one also slows down codegen, I think it might help to use direct `match` instead of all those uses of `as_ref()`, `as_mut()`, and `?`.

File: src/libcore/iter/adapters/fuse.rs [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+}
+
+// NOTE: for `I: FusedIterator`, we assume that the iterator is always `Some`
+impl<I: FusedIterator> Fuse<I> {
+    #[inline(always)]
+    fn as_inner(&self) -> &I {
+        match self.iter {
+            Some(ref iter) => iter,
+            // SAFETY: the specialized iterator never sets `None`
+            None => unsafe { intrinsics::unreachable() },
```
Comment (Mark-Simulacrum) [inline src/libcore/iter/adapters/fuse.rs]: Ideally all of these should be `hint::unreachable_unchecked()` today I think

Comment (cuviper) [reply src/libcore/iter/adapters/fuse.rs]: Why is that better? For `core` that's free to use intrinsics, going through `hint` just adds more inline indirection...

Comment (Mark-Simulacrum) [reply src/libcore/iter/adapters/fuse.rs]: Clearer meaning, IMO. But not that important, certainly no need to change things in this PR.

**Comment by timvermeulen** [comment]
Does this affect when drop handlers are called? (I don't know how important this is)

**Comment by scottmcm** [comment]
@timvermeulen Yes it does.

@rust-lang/libs I was going to r+ this, as it's a size win and seems fine in perf.  But now that #70374 showed up too, I wanted to ask how you feel about things.  (And to double-check that where `Drop` happens isn't something that needs to be preserved.)

File: src/libcore/iter/adapters/fuse.rs (hunk 1 above)
Comment (RalfJung) [reply src/libcore/iter/adapters/fuse.rs]: > Why is that better?

For example, if we make the `hint` method panic in debug-builds at some point (to detect incorrect use), then using the intrinsic would not benefit from that check.

**Review by dtolnay — CHANGES_REQUESTED** [review]
This PR drops the impl TrustedRandomAccess for Fuse\<I\> if I is !FusedIterator, which seems unintentional.

```diff
  unsafe impl<I> TrustedRandomAccess for Fuse<I>
  where
-     I: TrustedRandomAccess,
+     I: TrustedRandomAccess + FusedIterator,
```

Otherwise looks good to me.

**Comment by cuviper** [comment]
@dtolnay that was semi-intentional, because what is that supposed to do if we've already set `None`? The old code just ignored `done` and called the iterator regardless.

But `TrustedRandomAccess` is an internal trait anyway, and all who implement that are already `FusedIterator` too. Maybe we should just formalize that?

**Comment by dtolnay** [comment]
The caller of get_unchecked guarantees that the fuse isn't already None, because they have a specific index at which they know an element exists. So you should be able to use the same approach as in as_inner_mut.

**Comment by cuviper** [comment]
@dtolnay OK, done!

**Review by dtolnay — APPROVED** [review]

**Comment by AnthonyMikh** [comment]
@dtolnay since #70374 modifes `Fuse` too, may you take a look at that PR as well?

**Comment by scottmcm** [comment]
Thanks for taking this on, BTW, @cuviper!

File: src/libcore/iter/adapters/fuse.rs (hunk 1 above)
Comment (AnthonyMikh) [reply src/libcore/iter/adapters/fuse.rs]: One should note that this code triggers UB if an iterator implements `FusedIterator` by error. The implementor don't even have to use `unsafe` in order to achieve that.

Comment (dtolnay) [reply src/libcore/iter/adapters/fuse.rs]: > One should note that this code triggers UB if an iterator implements `FusedIterator` by error. The implementor don't even have to use `unsafe` in order to achieve that.

I don't think so. Regardless of `I`'s FusedIterator implementation or anything in `I`, *it* can't set the `iter` field of `Fuse` to None. This modules never sets `iter` to None if `I` is FusedIterator.

**Pull request merged**